### PR TITLE
Add fast beam search emformer

### PR DESCRIPTION
**Pull request by ezerhouni**
This PR adds ```fast_beam_search_one_best``` to the following models :

- [X] pruned_stateless_emformer_rnnt2

- [x] conv_emformer_transducer_stateless

### CODE REVIEW

File: sherpa/bin/pruned_stateless_emformer_rnnt2/decode.py [hunk 1]
```diff
@@ -15,8 +15,8 @@
 # limitations under the License.
 
 import math
-from typing import List
-
+from typing import List, Optional
+import k2
```
Comment (csukuangfj) [inline sherpa/bin/pruned_stateless_emformer_rnnt2/decode.py:19]: We are using [pep8](https://peps.python.org/pep-0008/#imports) style here:

```
Imports should be grouped in the following order:

Standard library imports.
Related third party imports.
Local application/library specific imports.

You should put a blank line between each group of imports.
```

And we use [isore](https://github.com/PyCQA/isort) to sort imports automatically

File: sherpa/bin/pruned_stateless_emformer_rnnt2/decode.py [hunk 2]
```diff
@@ -141,8 +148,24 @@ def __init__(
...
+            self.rnnt_decoding_stream = k2.create_rnnt_decoding_stream(
+                decoding_graph.arcs
+            )
+            self.hyp = []
+        elif decoding_method == "greedy_search":
+            assert decoder_out is not None
+            self.decoder_out = decoder_out
+            self.hyp = [blank_id] * context_size
+        else:
+            assert (
```
Comment (csukuangfj) [inline sherpa/bin/pruned_stateless_emformer_rnnt2/decode.py:164]: `assert False` throws an `AssertionError` exception.
I suggest
```python3
raise ValueError(f''Decoding method {decoding_method} is not supported.")
```
It expresses the intention directly.

File: sherpa/bin/pruned_stateless_emformer_rnnt2/streaming_server.py [hunk 3]
```diff
@@ -152,27 +191,40 @@ def run_model_and_do_greedy_search(
...
 
     state_list = []
-    decoder_out_list = []
-    hyp_list = []
     feature_list = []
+    processed_frames_list = []
+    if decoding_method == "greedy_search":
+        decoder_out_list = []
+        hyp_list = []
+    else:
```
Comment (csukuangfj) [inline sherpa/bin/pruned_stateless_emformer_rnnt2/streaming_server.py:204]: Please refactor it in a way that makes it easier to support other decoding methods.

File: sherpa/bin/pruned_stateless_emformer_rnnt2/streaming_server.py [hunk 4]
```diff
@@ -131,11 +136,45 @@ def get_args():
...
+            fast_beam_search are implemented.
+            """,
+    )
+
     return parser.parse_args()
 
 
 @torch.no_grad()
-def run_model_and_do_greedy_search(
+def run_model_and_do_search(
```
Comment (csukuangfj) [inline sherpa/bin/pruned_stateless_emformer_rnnt2/streaming_server.py:177]: I suggest refactoring it into two functions:
- run_model_and_do_greedy_search
- run_model_and_do_fast_beam_search

It makes the code more readable and more maintainable, I think.

If we want to support another decoding method, we can just create a new function to do that.

----

It is just a suggestion. You can leave it as it is.

File: sherpa/bin/pruned_stateless_emformer_rnnt2/decode.py (hunk 1 above)
Comment (ezerhouni) [reply sherpa/bin/pruned_stateless_emformer_rnnt2/decode.py:19]: Usually the pre-commit hook was taking care of it, that's why I notice that there are none. Will fix it

File: sherpa/bin/pruned_stateless_emformer_rnnt2/streaming_server.py (hunk 4 above)
Comment (ezerhouni) [reply sherpa/bin/pruned_stateless_emformer_rnnt2/streaming_server.py:177]: It make sense, I will finish first the work of conv_emformer at first

File: sherpa/bin/conv_emformer_transducer_stateless/decode.py [hunk 5]
```diff
@@ -1,5 +1,4 @@
-# Copyright      2022  Xiaomi Corp.        (authors: Fangjun Kuang
-#                                                    Zengwei Yao)
```
Comment (csukuangfj) [inline sherpa/bin/conv_emformer_transducer_stateless/decode.py:2]: Please keep the authors info.

File: sherpa/bin/pruned_stateless_emformer_rnnt2/decode.py (hunk 1 above)
Comment (ezerhouni) [reply sherpa/bin/pruned_stateless_emformer_rnnt2/decode.py:20]: It should be fine by now

File: sherpa/bin/pruned_stateless_emformer_rnnt2/decode.py (hunk 2 above)
Comment (ezerhouni) [reply sherpa/bin/pruned_stateless_emformer_rnnt2/decode.py]: Done

**Comment by csukuangfj** [comment]
Could you also update GitHub CI for https://github.com/k2-fsa/sherpa/blob/master/.github/workflows/run-streaming-conv-emformer-test.yaml#L36

to include
https://github.com/k2-fsa/sherpa/blob/fb0ee4adfdd5e490111ada244e0713410561d2dd/.github/workflows/run-streaming-conformer-test.yaml#L37

https://github.com/k2-fsa/sherpa/blob/fb0ee4adfdd5e490111ada244e0713410561d2dd/.github/workflows/run-streaming-conformer-test.yaml#L127

You can apply a label to this PR with the name `ready` to trigger GitHub actions.

File: sherpa/bin/conv_emformer_transducer_stateless/decode.py (hunk 5 above)
Comment (ezerhouni) [reply sherpa/bin/conv_emformer_transducer_stateless/decode.py:2]: Oups my bad. Fixed

File: sherpa/csrc/rnnt_emformer_model.cc [hunk 6]
```diff
@@ -53,7 +53,7 @@ RnntEmformerModel::RnntEmformerModel(const std::string &filename,
   right_context_length_ = encoder_.attr("right_context_length").toInt();
 }
 
-std::pair<torch::Tensor, RnntEmformerModel::State>
+std::tuple<torch::Tensor, torch::Tensor, RnntEmformerModel::State>
```
Comment (csukuangfj) [inline sherpa/csrc/rnnt_emformer_model.cc:56]: For the following CI error
https://github.com/k2-fsa/sherpa/runs/7444603197?check_suite_focus=true

You can check the style of the C++ code locally by
```bash
./scripts/check_style_cpplint.sh 2
```

Comment (ezerhouni) [reply sherpa/csrc/rnnt_emformer_model.cc:56]: Yup, already working on it ;)

**Comment by ezerhouni** [comment]
Should be all good @csukuangfj !

**Review by csukuangfj — APPROVED** [review]

**Comment by csukuangfj** [comment]
> Should be all good @csukuangfj !

Thanks! Let us wait for the CI.  Will merge after all tests pass. 

Thanks again!
